code-analysis/crates/codegen/parser/src/chumsky/code_generator.rs

```diff
@@ -0,0 +1,78 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use inflector::Inflector;
+use proc_macro2::{Ident, TokenStream};
+use quote::format_ident;
+use semver::Version;
+
+use super::naming;
+
+#[derive(Clone, Debug, Default)]
+pub struct CodeGenerator {
+    pub token_kinds: BTreeMap<String, Option<String>>,
+    pub rule_kinds: BTreeSet<String>,
+    pub parsers: BTreeMap<String, GeneratedParser>,
+    pub errors: Vec<String>,
+}
+
+#[derive(Clone, Debug, Default)]
+pub enum ParserResultType {
+    #[default]
+    Token,
+    Rule,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct GeneratedParser {
+    pub comment: Vec<String>,
+    pub result_type: ParserResultType,
+    pub versions: BTreeMap<Version, TokenStream>,
+}
+
+impl CodeGenerator {
+    pub fn add_rule_kind(&mut self, name: String) -> Ident {
+        let kind = name.to_pascal_case();
+        let ident = format_ident!("{}", kind);
+        self.rule_kinds.insert(kind);
+        ident
+    }
+
+    pub fn add_token_kind(&mut self, name: String) -> Ident {
+        let kind = name.to_pascal_case();
+        let ident = format_ident!("{}", kind);
+        self.token_kinds.insert(kind, None);
+        ident
+    }
+
+    pub fn add_terminal_kind(&mut self, terminal: String) -> Ident {
+        let kind = naming::name_of_terminal_string(&terminal).to_pascal_case();
+        let ident = format_ident!("{}", kind);
+        self.token_kinds.insert(kind, Some(terminal));
+        ident
+    }
+
+    pub fn add_parser(
+        &mut self,
+        name: String,
+        version: &Version,
+        comment: Vec<String>,
+        body: TokenStream,
+        result_type: ParserResultType,
+    ) {
+        // TODO: assert consistency of return types across all versions of the same
```

Comment:
nit: maybe an issue for this (and similar TODOs) instead of a comment? similar to #181 so that we can track it.

Also, if possible, I suggest adding such validation to the main `schema` crate, that is used by other codegen pieces. They should not be able to build successfully either if the input grammar is invalid.

Reply:
I will add an issue and audit the code for more instances.

As previously discussed, some validation requires that you have done some of the code-gen work to detect the errors, so I think a pure waterfall validate-before-anything isn't going to be feasible.